### PR TITLE
[RedSuns Review] Add progress visualization feature

### PR DESCRIPTION
**Pull request by kkawa14**
## Pull Request Description:
- Add progress visualization feature.
  - Modify the calling functions in PTQ/GPTQ and Pytorch/Keras so they call module progress_info_controller .
  - Add new DebugConfig Class member variable.
- Add unit test for module progress_info_controller and e2e test. 

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/quantization/debug_config.py
Comment (yt0705) [inline model_compression_toolkit/core/common/quantization/debug_config.py:32]: >        simulate_scheduler (bool): Simulate scheduler behavior to compute operators' order and cuts.
>        bypass (bool): A flag to enable MCT bypass, which skips MCT runner and returns the input model unchanged.

The description for "progress_info_callback" is missing.

File: model_compression_toolkit/core/common/progress_config/progress_info_controller.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        """Close progress bar."""
+        if self.pbar is not None:
+            self.pbar.close()
+            self.pbar = None
+
+
+def research_progress_total(
+    core_config: 'CoreConfig',
+    target_resource_utilization: 'ResourceUtilization' = None,
+    gptq_config: 'GradientPTQConfig' = None,
```
Comment (yt0705) [inline model_compression_toolkit/core/common/progress_config/progress_info_controller.py]: The apostrophe is unnecessary in type hints.

ex.
 core_config: 'CoreConfig' -> core_config: CoreConfig
 gptq_config: 'GradientPTQConfig' -> gptq_config: GradientPTQConfig

File: model_compression_toolkit/core/common/progress_config/progress_info_controller.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+) -> int:
```
Comment (yt0705) [inline model_compression_toolkit/core/common/progress_config/progress_info_controller.py]: Align the paragraph with other functions

```
def research_progress_total(core_config: CoreConfig,
                            target_resource_utilization: ResourceUtilization = None,
                            gptq_config: GradientPTQConfig = None,
                            ) -> int:
```

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py [hunk 3]
```diff
@@ -183,6 +189,10 @@ def ensure_maxbit_minimal_metric(node_candidates_metrics, max_ind):
             metrics[max_ind] = max_val
             return metrics
 
+        position = int(self.progress_info_controller is not None)
```
Comment (yt0705) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py]: > position = int(self.progress_info_controller is not None)

What does this variable mean?
It appears to be unreferenced.

File: model_compression_toolkit/gptq/runner.py
Comment (yt0705) [inline model_compression_toolkit/gptq/runner.py:60]: >        fw_impl: Framework implementation per framework
>        hessian_info_service: HessianInfoService to fetch information based on the hessian approximation for the float model.

The description comment for progress_info_controller is missing.

Comment (yt0705) [inline model_compression_toolkit/gptq/runner.py:103]: >        tb_w: A TensorBoardWriter object initialized with the logger dir path if it was set, or None otherwise.
>        hessian_info_service: HessianScoresService to fetch approximations of the hessian scores for the float model.

The description comment for progress_info_controller is missing.

**Review by yt0705 — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/common/quantization/debug_config.py
Comment (kkawa14) [reply model_compression_toolkit/core/common/quantization/debug_config.py:32]: The argument description for "progress_info_callback" has been documented.

File: model_compression_toolkit/core/common/progress_config/progress_info_controller.py (hunk 2 above)
Comment (kkawa14) [reply model_compression_toolkit/core/common/progress_config/progress_info_controller.py]: I’ve aligned it.

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py (hunk 3 above)
Comment (kkawa14) [reply model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_manager.py]: It’s something I forgot to delete from when I was working on a different approach.
So I removed it.

File: model_compression_toolkit/gptq/runner.py
Comment (kkawa14) [reply model_compression_toolkit/gptq/runner.py:60]: I have added the missing description of progress_info_controller.

Comment (kkawa14) [reply model_compression_toolkit/gptq/runner.py:103]: I have added the missing description of progress_info_controller.

File: model_compression_toolkit/core/common/progress_config/progress_info_controller.py (hunk 1 above)
Comment (kkawa14) [reply model_compression_toolkit/core/common/progress_config/progress_info_controller.py]: Because a circular import error occurs, it’s not possible to apply the changes as you suggested.
Therefore, this statement is written to be used in conjunction with "TYPE_CHECKING".
[NodeWeightsQuantizationConfig Class](https://github.com/SonySemiconductorSolutions/mct-model-optimization/blob/main/model_compression_toolkit/core/common/quantization/node_quantization_config.py#L396) also terminates with the same error, so it is avoided by using "TYPE_CHECKING".

File: tests_pytest/pytorch_tests/e2e_tests/test_progress_visualization_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import pytest
+
+import model_compression_toolkit as mct
+from model_compression_toolkit.gptq import pytorch_gradient_post_training_quantization
+from model_compression_toolkit.ptq import pytorch_post_training_quantization
```
Comment (kawakami-masaki0) [inline tests_pytest/pytorch_tests/e2e_tests/test_progress_visualization_pytorch.py]: Please remove imported but unused modules.

File: tests_pytest/keras_tests/e2e_tests/test_progress_visualization_keras.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import pytest
+
+import model_compression_toolkit as mct
+from model_compression_toolkit.gptq import keras_gradient_post_training_quantization
+from model_compression_toolkit.ptq import keras_post_training_quantization
```
Comment (kawakami-masaki0) [inline tests_pytest/keras_tests/e2e_tests/test_progress_visualization_keras.py]: Please remove imported but unused modules.

File: tests_pytest/common_tests/unit_tests/core/progress_config/test_research_progress_total.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#  ==============================================================================
+
+import pytest
+from unittest.mock import Mock
+
+from dataclasses import dataclass
```
Comment (kawakami-masaki0) [inline tests_pytest/common_tests/unit_tests/core/progress_config/test_research_progress_total.py]: Please remove imported but unused modules.

File: tests_pytest/common_tests/unit_tests/core/progress_config/test_progress_control_module.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#  ==============================================================================
+
+import pytest
+from unittest.mock import Mock
```
Comment (kawakami-masaki0) [inline tests_pytest/common_tests/unit_tests/core/progress_config/test_progress_control_module.py]: Please remove imported but unused modules.

File: tests_pytest/keras_tests/e2e_tests/test_progress_visualization_keras.py (hunk 5 above)
Comment (kkawa14) [reply tests_pytest/keras_tests/e2e_tests/test_progress_visualization_keras.py]: All of the libraries imported here are used.

File: tests_pytest/pytorch_tests/e2e_tests/test_progress_visualization_pytorch.py (hunk 4 above)
Comment (kkawa14) [reply tests_pytest/pytorch_tests/e2e_tests/test_progress_visualization_pytorch.py]: All of the libraries imported here are used.

File: tests_pytest/common_tests/unit_tests/core/progress_config/test_research_progress_total.py (hunk 6 above)
Comment (kkawa14) [reply tests_pytest/common_tests/unit_tests/core/progress_config/test_research_progress_total.py]: I removed "dataclasses" modules.

File: tests_pytest/common_tests/unit_tests/core/progress_config/test_progress_control_module.py (hunk 7 above)
Comment (kkawa14) [reply tests_pytest/common_tests/unit_tests/core/progress_config/test_progress_control_module.py]: I removed "unittest.mock" modules.

File: tests_pytest/pytorch_tests/e2e_tests/test_progress_visualization_pytorch.py (hunk 4 above)
Comment (kkawa14) [reply tests_pytest/pytorch_tests/e2e_tests/test_progress_visualization_pytorch.py]: Sorry, I understand what you mean now.
Indeed, the imported module here is not used.
So, I removed it.

File: tests_pytest/keras_tests/e2e_tests/test_progress_visualization_keras.py (hunk 5 above)
Comment (kkawa14) [reply tests_pytest/keras_tests/e2e_tests/test_progress_visualization_keras.py]: Sorry, I understand what you mean now.
Indeed, the imported module here is not used.
So, I removed it.
